src/MumbleUDP.proto

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+syntax = "proto3";
+
+package MumbleUDP;
+
+option optimize_for = SPEED;
+
+message Audio {
+	oneof Header {
+		// When this audio is sent by the client to the server, this is set to the target of the audio data. This target
+		// is a number in the range [0, 2^{32} - 1], where 0 means "normal talking", 2^{5} - 1 means "server loopback"
+		// and all other targets are understood as shout/whisper targets that have previously been registered via a
+		// VoiceTarget message (via TCP).
+		uint32 target  = 1;	
+		// When this audio is sent by the server to the client, this indicates the context in which the audio has been sent.
+		// 0: Normal speech
+		// 1: Shout to channel
+		// 2: Whisper to user
+		// 3: Received via channel listener
+		uint32 context = 2;
+	};
+
+	// The session of the client (sender) this audio was originally sent from. This field is not required when sending
+	// audio to the server, but will always be set when receiving audio from the server.
+	uint32 sender_session = 3;
+
+	// The number of the first contained audio frame (indicating the position of that frame in the overall audio stream)
+	uint64 frame_number = 4;
+
+	// The actual voice data payload in the Opus format.
+	bytes opus_data = 5;
+
+	// Optional positional data indicating the speaker's position in a virtual world (in meters). This "list" is really
+	// expected to be an array of size 3 containing the X, Y and Z coordinates of the position (in that order).
+	repeated float positional_data = 6;
+
+	// A volume adjustment determined by the server for this audio packet. It is up to the client to apply this adjustment to
+	// the resulting audio (or not). Note: A value of 0 means that this field is unset.
+	float volume_adjustment = 7;
+
+	// A flag indicating whether this audio packet represents the end of transmission for the current audio stream
+	bool is_terminator = 16;
```

Comment:
Because the terminator is not part of every packet (it is only sent at the end of a stream). Thus we can reserve the lower-index fields (which only require 1 byte encoding overhead) for future fields that will be needed more frequently and accept the overhead of having to use a second byte encoding overhead for fields like the terminator that are only sent rarely (comparatively)

Reply:
Shouldn't the 1 byte encoding overhead occur at index 256 (0xFF + 1)?

Reply:
No, because the encoding overhead contains the field index as well as the field type. The details are given at https://developers.google.com/protocol-buffers/docs/encoding but essentially, Protobuf uses varint encoding, so you have 7 bits to represent a one-byte number (the 8th one is the continuation flag for the varint encoding). Three bits are used to encode the field type and the remaining 4 bits are used to encode the field index (2^4 = 16 -> Thus the values 0-15 can be encoded in a single byte of overhead)